letshelp-letsencrypt/letshelp_letsencrypt/letshelp_letsencrypt_apache.py

```diff
@@ -0,0 +1,303 @@
+
+
+def safe_config_file(config_file):
+    """Returns True if config_file can be safely copied
+
+    :param str config_file: Path to an Apache configuration file
+
+    :returns: True if config_file can be safely copied
+    :rtype: bool
+
+    """
+    config_file_lower = config_file.lower()
+    if _SENSITIVE_FILENAME_REGEX.search(config_file_lower):
+        return False
+
+    proc = subprocess.Popen(["file", config_file],
+                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    file_output, _ = proc.communicate()
+
+    if "ASCII" in file_output:
+        possible_password_file = empty_or_all_comments = True
+        with open(config_file) as config_fd:
+            for line in config_fd:
+                if not (line.isspace() or line.lstrip().startswith("#")):
+                    empty_or_all_comments = False
+                    if line.startswith("-----BEGIN"):
+                        return False
+                    elif not ":" in line:
+                        possible_password_file = False
+        # If file isn't empty or commented out and could be a password file,
+        # don't include it in selection. It is safe to include the file if
+        # it consists solely of comments because comments are removed before
+        # submission.
+        return empty_or_all_comments or not possible_password_file
+
+    return False
+
+
+def setup_tempdir(args):
+    """Creates a temporary directory and necessary files for config
+
+    :param argparse.Namespace args: Parsed command line arguments
+
+    :returns: Path to temporary directory
+    :rtype: str
+
+    """
+    tempdir = tempfile.mkdtemp()
+
+    with open(os.path.join(tempdir, "config_file"), "w") as config_fd:
+        config_fd.write(args.config_file + "\n")
+
+    proc = subprocess.Popen([args.apache_ctl, "-v"],
+                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    with open(os.path.join(tempdir, "version"), "w") as version_fd:
+        version_fd.write(proc.communicate()[0])
+
+    proc = subprocess.Popen([args.apache_ctl, "-d", args.server_root, "-f",
+                             args.config_file, "-M"],
+                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    with open(os.path.join(tempdir, "modules"), "w") as modules_fd:
+        modules_fd.write(proc.communicate()[0])
+
+    proc = subprocess.Popen([args.apache_ctl, "-d", args.server_root, "-f",
+                             args.config_file, "-t", "-D", "DUMP_VHOSTS"],
+                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    with open(os.path.join(tempdir, "vhosts"), "w") as vhosts_fd:
+        vhosts_fd.write(proc.communicate()[0])
+
+    return tempdir
+
+
+def verify_config(args):
+    """Verifies server_root and config_file specify a valid config
+
+    :param argparse.Namespace args: Parsed command line arguments
+
+    """
+    try:
+        with open(os.devnull, "w") as devnull:
```

Comment:
Meeeh, https://docs.python.org/3/library/subprocess.html#subprocess.DEVNULL, new in version 3.3 :(